### PR TITLE
LLVM codegen changes for SPIR backend

### PR DESCRIPTION
**Pull request by Solaryee**
This PR aims to add LLVM codegen changes for SPIRV backend. It contains SPIRV target and related intrinsic.

### CODE REVIEW

File: xla/service/gpu/parallel_loop_emitter.cc
Comment (akuegel) [inline xla/service/gpu/parallel_loop_emitter.cc:186]: As far as I understand, these changes are just a NFC refactoring, and are not related to the real changes being done in this PR? If yes, maybe split them into their own PR?
I would say the new code is not necessarily more easy to read. I agree with the changes to swap if/else blocks and get rid of the negation, but I would prefer if the loop still starts from 1 and we handle the i = 0 case outside the loop.

Comment (ShengYang1) [reply xla/service/gpu/parallel_loop_emitter.cc:186]: Yes, this is a NFC refactoring, I can put it in another PR.

**Review by sergeykozub — APPROVED** [review]

File: xla/service/gpu/parallel_loop_emitter.cc
Comment (ShengYang1) [reply xla/service/gpu/parallel_loop_emitter.cc]: Done. This PR only contains SPIR target/intrinsic now.

**Review by akuegel — APPROVED** [review]

**Review by cheshire — CHANGES_REQUESTED** [review]
Is it possible to test those changes? How do we plan to trigger Intel codegen codepath?

File: xla/service/gpu/target_util.cc
Comment (penpornk) [inline xla/service/gpu/target_util.cc:65]: I think this can be refactored to a templated struct because we only need the code for one device backend at the same time. But I can look into that after this PR.

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR!

**Comment by penpornk** [comment]
Re: @cheshire:
> Is it possible to test those changes? How do we plan to trigger Intel codegen codepath?

@ShengYang1 Do you already have a CI running? Are the results publicly available now? If so, could you please share the CI link here for the time being? (We can look into adding the CI link to our github check dashboard later.)

File: xla/service/gpu/target_util.cc
Comment (penpornk) [inline xla/service/gpu/target_util.cc:48]:
```suggestion
  std::variant<llvm::Intrinsic::ID,
```

File: xla/service/gpu/target_util.cc
Comment (penpornk) [inline xla/service/gpu/target_util.cc:390]:
```suggestion
        std::get_if<llvm::Intrinsic::ID>(
```

File: xla/service/gpu/target_util.cc
Comment (penpornk) [inline xla/service/gpu/target_util.cc:396]:
```suggestion
          std::get_if<std::function<llvm::CallInst*(llvm::IRBuilder<>*)>>(
```

**Review by penpornk — CHANGES_REQUESTED** [review]
More feedback from our code check. (For these functions, we prefer the `std` version.)

**Comment by Zantares** [comment]
> Re: @cheshire:
> 
> > Is it possible to test those changes? How do we plan to trigger Intel codegen codepath?
> 
> @ShengYang1 Do you already have a CI running? Are the results publicly available now? If so, could you please share the CI link here for the time being? (We can look into adding the CI link to our github check dashboard later.)

According to the [suggestion ](https://github.com/openxla/community/pull/99#issuecomment-1859630963) in RFC, we have initial CI plan here and welcome your suggestions:
1. We plan to enable post-CI as AMD, but we also would like to support more capabilities (trigger GitHub action on demand) if OpenXLA is interested :)
2. Current public Intel CI can't be used to track OpenXLA main branch because it based on Extension with fixed OpenXLA commit. Consider the rebase effort, we hope to fully enable the CI after Runtime PR is merge (it's the largest and most important PR), then we can directly build Intel GPU target and check it without Extension. Furthermore, make these pending PRs waiting till Runtime PR is submitted, and test them together is an alternative solution 
3. We will make sure these PR won't impact existing target (check by current OpenXLA pre-CI)

**Review by tdanyluk — APPROVED** [review]

**Review by penpornk — APPROVED** [review]
Thank you for the changes!
@cheshire Are you okay with us going ahead and merging the PR now?

**Comment by cheshire** [comment]
I think we should not merge this just yet; without tests and an integration plan the diff by itself makes little sense.

Is there a working branch where all this code works end-to-end? How many tests are passing now? What is the desired end state? What are the CI plans?

**Comment by Zantares** [comment]
Hi @cheshire , hope below info can answer your question:
1. The integration plan is listed in RFC https://github.com/openxla/community/blob/49541f1e8b0bdba6ca003797a5ae01f4a2f8cbb7/rfcs/20231102-intel-gpu.md#overview. The basic functionality can be enabled with PR1 & PR3, that's why we suggest testing it after these 2 PRs are merged.
    - SPIRV target (this PR)
    - OneDNN lib call. This part will refer the 3rd library call in CPU RFC https://github.com/openxla/community/pull/96, so it will be deferred after the related CPU changes are ready
    - SYCL runtime (To be submitted)

2. The CI plan is also roughly listed in RFC https://github.com/openxla/community/blob/49541f1e8b0bdba6ca003797a5ae01f4a2f8cbb7/rfcs/20231102-intel-gpu.md#engineering-impact. Furthermore, we have enabled such public CI based on [Intel® Extension for OpenXLA*](https://github.com/intel/intel-extension-for-openxla) and shared it in [OpenXLA community meeting](https://www.youtube.com/watch?v=YK1CLzIcsJ8&list=PLlFotmaRrOzu8TQsTahDo_Cn7QdntFlUL&index=1&t=509s). Right now, it can only work with specific OpenXLA commit and Intel Extension. Once the related PRs are merged, it can work with OpenXLA main branch.
3. `Is there a working branch where all this code works end-to-end? How many tests are passing now?`
Yes, as I mentioned above we have all these changes in specific OpenXLA commit (worked with Intel Extension) and we can pass 95 JAX native UT from [public JAX repo](https://github.com/google/jax/tree/main/tests) and [several popular models here](https://github.com/intel/intel-extension-for-openxla/tree/0.2.0/example)
4. `What is the desired end state?`
Any hardware based on generic SPIRV/SYCL can be functionality enabled after all PRs are merged. To get advanced performance, plug-in/extension are still needed for critical operations.

**Comment by cheshire** [comment]
OK!

File: xla/service/gpu/target_util.cc
Comment (cheshire) [reply xla/service/gpu/target_util.cc:65]: +1, there's quite a bit of duplication across those `EmitDeviceFunctionCall`, maybe just extracting it to a function would be better?

File: xla/service/gpu/target_util.cc
Comment (cheshire) [inline xla/service/gpu/target_util.cc:318]: Nit: XLA-preferred style is to use braces everywhere, even for single lines.

**Review by cheshire — CHANGES_REQUESTED** [review]
Two nits.

File: xla/service/gpu/target_util.cc
Comment (ShengYang1) [reply xla/service/gpu/target_util.cc:318]: Done.

**Review by akuegel — CHANGES_REQUESTED** [review]

File: xla/service/gpu/ir_emission_utils.cc
Comment (ShengYang1) [reply xla/service/gpu/ir_emission_utils.cc]: Done.

**Review by akuegel — APPROVED** [review]
Thank you for the quick fix :)

File: xla/service/gpu/target_util.cc
Comment (penpornk) [reply xla/service/gpu/target_util.cc:65]: That would make it cleaner too :)

**Pull request closed without merging**
